File: apps/web/lib/moneda.ts
/**
 * BUG-092 — la única frontera donde un importe se convierte en texto.
 *
 * ## Qué había
 *
 * `tenant.settings.currency` ofrecía MXN, USD y EUR y **el formulario que la
 * guardaba era el único sitio que la leía**. Las diez superficies que muestran
 * dinero traían `currency: "MXN"` escrito a mano, así que un inquilino en
 * dólares —el propio sembrado crea uno— veía sus importes **rotulados en
 * pesos**. El número no estaba mal; la unidad era mentira, que en un importe es
 * lo mismo.
 *
 * ## La decisión del owner (2026-08-07)
 *
 * La moneda va **sobre el proyecto**. El inquilino declara una *preferida*, que
 * es el valor inicial de los proyectos que no eligen una propia. Cada
 * superficie tiene que reflejar la que corresponde.
 *
 * ## Por qué `moneda` no tiene valor por defecto
 *
 * Misma lección que `confirmar.ts` y que `MarcaDeDatos`: **un parámetro con
 * defecto es un parámetro que nadie rellena.** Si `formatearImporte` asumiera
 * MXN cuando no le pasan nada, el bug volvería exactamente igual y en el mismo
 * sitio — la diferencia entre `currency: "MXN"` escrito a mano y un defecto
 * escondido en una función es de dónde está escrito, no de qué hace.
 *
 * El código de moneda lo trae la API, ya resuelto: el frontend nunca aplica la
 * regla «nulo significa la del inquilino», porque esa regla vive en
 * `app/dominio/moneda.py` y dos sitios decidiendo lo mismo divergen.
 */

/** Los tres códigos que el producto admite. */
export const MONEDAS = ["MXN", "USD", "EUR"] as const;
export type Moneda = (typeof MONEDAS)[number];

/**
 * El último recurso cuando ni el proyecto ni el inquilino dicen nada. Es la que
 * el producto venía aplicando de facto, así que ningún importe existente cambia
 * de rótulo. Espeja `dominio.moneda.POR_DEFECTO`, y una prueba lo comprueba.
 */
export const POR_DEFECTO = "MXN";

/** El locale con el que se agrupan los millares. No es la moneda. */
const LOCALE = "es-MX";

/**
 * Un importe con su moneda. `moneda` es obligatoria y no tiene defecto.
 *
 * `null` y `undefined` devuelven el marcador de «sin dato» (DAT-12): un
 * presupuesto no cargado no es un presupuesto de cero, y pintarlo como `$0`
 * dice que alguien decidió no gastar cuando lo que pasa es que nadie lo
 * capturó.
 */
export function formatearImporte(
  valor: number | null | undefined,
  moneda: string,
  opciones?: { decimales?: number },
): string {
  if (valor === null || valor === undefined || Number.isNaN(valor)) return "—";
  return new Intl.NumberFormat(LOCALE, {
    style: "currency",
    currency: moneda,
    maximumFractionDigits: opciones?.decimales ?? 0,
  }).format(valor);
}

/**
 * Un desglose `{moneda: importe}` como texto, para los agregados de cartera.
 *
 * **Con varias monedas no hay un total**, y esta función no lo inventa: las
 * pinta todas separadas. Sumar 1.000 MXN y 1.000 EUR para escribir «2.000» es
 * producir un número que no existe en ninguna parte, y convertir exigiría un
 * tipo de cambio con fecha y con firma — que es otro frente.
 *
 * Sin ninguna devuelve el marcador de «sin dato», que es distinto de cero por
 * el mismo motivo de siempre.
 */
export function formatearDesglose(porMoneda: Record<string, number>): string {
  const entradas = Object.entries(porMoneda ?? {});
  if (entradas.length === 0) return "—";
  return entradas
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([moneda, valor]) => formatearImporte(valor, moneda))
    .join(" · ");
}

/** ¿El desglose está en una sola moneda? Decide si se pinta uno o varios. */
export function monedaUnica(porMoneda: Record<string, number>): string | null {
  const codigos = Object.keys(porMoneda ?? {});
  return codigos.length === 1 ? codigos[0] : null;
}

File: apps/api/app/dominio/moneda.py
"""La moneda de un importe: del proyecto, con la del inquilino como preferida.

> **Decisión del owner, 2026-08-07 (BUG-092).** «Habrá una moneda preferida,
> pero la moneda irá sobre el proyecto y deben poder escoger la que necesite el
> proyecto. Debe reflejar la correspondiente.»

## Qué había antes

`tenant.settings.currency` ofrecía MXN, USD y EUR, y **el formulario que la
guardaba era el único sitio que la leía**. Las diez superficies que muestran
dinero traían `currency: "MXN"` escrito a mano, así que un inquilino en dólares
—el propio sembrado crea uno— veía sus importes **rotulados en pesos**. No es
que el número estuviera mal: es que la unidad era mentira, que en un importe es
lo mismo que estar mal.

Salió midiendo DAT-02, y DAT-01 lo dejó declarado con su disparador: «la unidad
canónica del importe pasa a ser la moneda del inquilino el día que llegue a la
presentación». El owner movió el grano un escalón más abajo, al proyecto, que es
donde de verdad vive un presupuesto.

## La parte que no es evidente: agregar

Un portafolio con un proyecto en pesos y otro en euros **no tiene un
presupuesto total**. Sumar 1.000 MXN y 1.000 EUR y escribir «2.000» es inventar
un número que no existe en ninguna parte, y es el error que este módulo existe
para hacer imposible: `agregar` devuelve **un importe por moneda**, nunca uno
solo.

Convertir tampoco es la salida: haría falta un tipo de cambio con fecha, y en
ese momento el número deja de ser un dato para ser una estimación — con todo lo
que eso arrastra (qué fuente, de qué día, quién la firma). Es un frente propio
y no se abre aquí.
"""
from __future__ import annotations

from collections.abc import Iterable
from decimal import Decimal

#: Las tres que el formulario de ajustes ya ofrecía. Se declaran aquí para que
#: el desplegable y la validación del servidor no puedan divergir — que es
#: exactamente lo que pasó con la moneda y la presentación.
MONEDAS: tuple[str, ...] = ("MXN", "USD", "EUR")

#: Lo que se aplica cuando ni el proyecto ni el inquilino dicen nada. Es la que
#: el producto venía usando de facto, así que ningún importe existente cambia de
#: rótulo al desplegar esto.
POR_DEFECTO = "MXN"


def es_valida(codigo: str | None) -> bool:
    return codigo in MONEDAS


def resolver(del_proyecto: str | None, preferida: str | None) -> str:
    """La moneda que se le aplica a un importe, en orden de especificidad.

    Proyecto → preferida del inquilino → `POR_DEFECTO`. El nulo del proyecto
    significa «la que diga el inquilino», no «ninguna»: así, cambiar la
    preferida arrastra a los proyectos que no eligieron, que es lo que espera
    quien la cambia.

    Un código desconocido se ignora en lugar de propagarse. `settings` lo edita
    una persona desde un formulario, y un importe sin rótulo por una errata es
    peor que uno con el rótulo por defecto.
    """
    if es_valida(del_proyecto):
        return str(del_proyecto)
    if es_valida(preferida):
        return str(preferida)
    return POR_DEFECTO


def agregar(importes: Iterable[tuple[str, Decimal | None]]) -> dict[str, Decimal]:
    """Suma importes **por moneda**. Nunca devuelve un total único.

    La firma es la mitad del control: no hay forma de llamar a esto y obtener un
    número solo. Un portafolio con un proyecto en pesos y otro en euros no tiene
    un presupuesto total, y escribir «2.000» donde hay 1.000 MXN y 1.000 EUR es
    inventar un número que no existe.

    Los nulos no suman y tampoco cuentan como cero: un proyecto sin presupuesto
    cargado no es un proyecto de presupuesto cero (DAT-12). Una moneda cuyos
    importes son todos nulos **no aparece** en el resultado.
    """
    por_moneda: dict[str, Decimal] = {}
    for codigo, importe in importes:
        if importe is None:
            continue
        moneda = codigo if es_valida(codigo) else POR_DEFECTO
        por_moneda[moneda] = por_moneda.get(moneda, Decimal(0)) + Decimal(importe)
    return por_moneda


def unica(por_moneda: dict[str, Decimal]) -> str | None:
    """El código si todos los importes están en la misma moneda, o `None`.

    Sirve para decidir la presentación: con una sola se pinta un número; con
    varias hay que pintarlas todas, y con ninguna no hay nada que pintar. Los
    tres casos son distintos y el llamador tiene que verlos distintos.
    """
    return next(iter(por_moneda)) if len(por_moneda) == 1 else None
